qa/tasks/cephfs/test_exports.py

```diff
@@ -157,3 +157,161 @@ def test_session_race(self):
+        return True
+
+    def _run_export_dir(self, importv, exportv):
+        # Wait till all MDS becomes active
+        self.fs.wait_for_daemons()
+
+        # Get all active ranks
+        ranks = self.fs.get_all_mds_rank()
+        if len(ranks) != 2:
+            log.error("Incorrect number of MDS ranks, exiting the test")
+            return False
+
+        rank_0 = self.fs.get_rank(rank=0)
+        rank_0_name = rank_0['name']
+
+        killpoint = {}
+        killpoint[0] = ('export', exportv)
+        killpoint[1] = ('import', importv)
+
+        command = ["config", "set", "mds_kill_export_at", str(exportv)]
+        result = self.fs.mds_asok(command, rank_0_id)
+        assert(result["success"])
+
+        command = ["config", "set", "mds_kill_import_at", str(importv)]
+        result = self.fs.mds_asok(command, rank_1_id)
+        assert(result["success"])
+
+        # This should kill either or both MDS process
+        command = ["export", "dir", "/abc", "1"]
+        try:
+            result = self.fs.mds_asok(command, rank_0_id)
+        except Exception as e:
+            log.error(e.__str__())
+
+        active_mds = self.fs.get_active_names()
+        if rank_0_name in active_mds:
+            log.info("MDS %s crashed at type %s killpoint %d"
+                    %(rank_0_name, killpoint[0][0], killpoint[0][1]))
+
+        # Waiting time for monitor to promote replacement for dead MDS
+        grace = int(float(self.fs.get_config("mds_beacon_grace", service_type="mon")))
+
+        #log_active_mds()
+        #log_crashed_mds()
+
+        time.sleep(grace * 2)
+
+        # Check if rank0 changed (standby tookover?)
+        new_rank_0 = self.fs.get_rank(rank=0)
+        self.assertEqual(rank_0['gid'], new_rank_0['gid'])
+        new_rank_0_name = new_rank_0['name']
+        active_mds = self.fs.get_active_names()
+        if new_rank_0_name in active_mds:
+            log.info("MDS %s active at type %s killpoint %d"
+                    %(new_rank_0_name, killpoint[0][0], killpoint[0][1]))
+
+
+        time.sleep(grace * 2)
+
+        active_mds = self.fs.get_active_names()
+        if len(active_mds) != 2:
+            "One or more MDS did not come up"
+            return False
+
+        if not self._verify_data():
+            return False;
+
+        return True
+
+
+    def _run_export(self, importv, exportv):
+        success_count = 0
+
+        for v in range(1, importv + 1):
+            for w in range(1, exportv + 1):
+                if self._run_export_dir(v, w):
+                    success_count = success_count + 1
+                else:
+                    log.error("Error for killpoint %d:%d" %(v, w))
+
+        for v in range(1, exportv + 1):
+            for w in range(1, importv + 1):
+                if self._run_export_dir(v, w):
+                    success_count = success_count + 1
+                else:
+                    log.error("Error for killpoint %d:%d" %(v, w))
```

Comment:
This kind of testing could be improved by testing specific things once and not iteratively. Bugs in one part of the code may influence other test cases.

It's a small amount of programming work to generate methods for each export/import pair. After defining `class TestKillPoints`, you can then use `setattr` to add a test method named e.g. `test_export_killpoint_1_4` that performs the export with `mds_kill_export_at=1` and `mds_kill_export_at=4`.

(Don't try to get fancy with python decorators or anything else to implement this.)

Reply:
Agree with making it modular and not doing it iteratively. Will change this.